mlflow/utils/autologging_utils.py

```diff
@@ -275,3 +282,201 @@ def batch_metrics_logger(run_id):
+    this metaclass wraps all class functions in a broad try / catch statement.
+    """
+
+    def __new__(cls, name, bases, dct):
+        for m in dct:
+            if hasattr(dct[m], "__call__"):
+                dct[m] = exception_safe_function(dct[m])
+        return type.__new__(cls, name, bases, dct)
+
+
+def _is_testing():
+    """
+    Indicates whether or not autologging functionality is running in test mode (as determined
+    by the `MLFLOW_AUTOLOGGING_TESTING` environment variable). Test mode performs additional
+    validation during autologging, including:
+        - Checks for the exception safety of arguments passed to model training functions
+          (i.e. all additional arguments should be "exception safe" functions or classes)
+        - Disables exception handling for patched function logic, ensuring that patch code
+          executes without errors during testing
+    """
+    import os
+
+    return os.environ.get("MLFLOW_AUTOLOGGING_TESTING", "false") == "true"
+
+
+def _validate_args(
+    user_call_args, user_call_kwargs, autologging_call_args, autologging_call_kwargs
+):
+    """
+    Used for testing purposes to verify that, when a patched model training function calls its
+    underlying / original training function, the following properties are satisfied:
+        - All arguments supplied to the patched model training function are forwarded
+          to the original training function
+        - Any additional arguments supplied to the original function are exception safe (i.e.
+          they are either functions decorated with the `@exception_safe_function` decorator
+          or classes / instances of classes with type `ExceptionSafeClass`
+    """
+
+    def _validate_new_arg(arg):
+        if type(arg) == list:
+            for item in arg:
+                _validate_new_arg(item)
+        elif callable(arg):
+            assert getattr(arg, _ATTRIBUTE_EXCEPTION_SAFE, False)
+        else:
+            import inspect
+
+            assert inspect.isclass(type(arg))
+            assert type(arg.__class__) == ExceptionSafeClass
+
+    def _validate(autologging_kwarg, user_kwarg=None):
+        if user_kwarg is None and autologging_kwarg is not None:
+            _validate_new_arg(autologging_kwarg)
+            return
+
+        assert type(autologging_kwarg) == type(user_kwarg)
+        if type(autologging_kwarg) == list:
+            user_kwarg = user_kwarg + ([None] * (len(autologging_kwarg) - len(user_kwarg)))
+            for a, u in zip(autologging_kwarg, user_kwarg):
+                _validate(a, u)
+        elif type(autologging_kwarg) == dict:
+            assert set(user_kwarg.keys()).issubset(set(autologging_kwarg.keys()))
+            for key in autologging_kwarg.keys():
+                _validate(autologging_kwarg[key], user_kwarg.get(key))
+        else:
+            assert autologging_kwarg is user_kwarg or autologging_kwarg == user_kwarg
```

Comment:
There are several assumptions encoded in this logic that may need to be revisited (they should at least be documented):
- Patch code doesn't make copies of objects that can't be compared via object equality
- Patch code never removes any user-supplied arguments / keyword arguments
- If patch code modifies user-supplied arguments / keyword arguments, modifications must be additive (e.g., appending an element to a list or a new key/value pair to a dictionary)